docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,256 @@
+
+1. Receive Message
+1. Start Delivery Process
+1. Attempt Individual Delivery
+
+Lets look into each of them.
+
+#### Receive Message
+
+* Start Transaction
+* Store Message
+* Commit Transaction
+* Send ACK
+* Trigger Delivery Process
+
+The acknowledgement transmission is intentionally not part of the transaction, as there is little bearing on the Producer actually receiving it. The worst case scenario will be Producer will resend the message.
+
+#### Start Delivery Process
+
+There is 2 entry points to the process.
+
+##### Triggered By Receipt Message
+
+* Retrieve the message
+* Retrieve the consumers
+* Start transaction
+* Create delivery status for each consumer
+* Mark message out-for-delivery
+* Commit transaction
+* Trigger independent asynchronous delivery process
+
+##### Triggered Fail-safe message
+
+* Retrieve all messages older than rational-delay
+* For each message use the Receipt Message trigger flow
+
+#### Attempt Individual Delivery
+
+For each message delivery we would have to follow the delivery lifecycle as designated by the _Message Delivery Status_. Here too, there will be 2 triggering functions -
+
+* Triggered by fail-safe mechanism
+  * Retrieve message deliveries in _Retry-Delivery_ with _next attempt timestamp_ past rational-delay
+  * For each message delivery run the _Triggered by Start Delivery Process_ process
+* Triggered by Start Delivery Process
+  * Update message delivery status to _Retry-In-Flight_ if the status is _Retry-Delivery_
+  * For delivery message retrieve consumer
+  * Attempt to deliver message
+  * Wait for the period of Delivery-Timeout
+    * If timed out then update status as _Retry-Delivery_ with exponentially backed off _next attempt timestamp_
+  * If work finishes before Timeout, mark status as _Delivered_
+  * If consumer failed to connect, update status as _Retry-Delivery_ with exponentially backed off _next attempt timestamp_
+  * If Max-Retries is reached and status is not _Delivered_ then mark delivery as _Dead_
+
+### Prioritized Dispatcher
+
+We can now overlay an aspect of _priority_ on this dispatcher process as well. For example, if there is a relatively higher priority message for a Consumer in a Channel, then in most cases that should be delivered first; one effect of introducing priority is, it will deliberately disrupt the order even further. Priority will allow producers to dictate in relative sense which should be delivered earlier provided they are both enqueued for delivery; it could very well be the cause that a relatively low and high priority message can be attempted to deliver concurrently. As the fail-safe mechanism will use the same process as the regular, it should respect priority if that is chosen. One important consideration would be, when priority is switched on, it will definitely impact overall throughput since there will be an active attempt to order queued messages by priority.
+
+## Use cases
```

Comment:
Reading through this spec, I am not quite sure which problem/use case has not already been addressed by RabbitMQ or SQS, and why we need yet another message broker.

And why we need an ESB specifically with HTTP interface (SQS talks HTTP as well though)? If the reason is to have easy to implement clients in major tech stacks, then again RabbitMQ and SQS have mature client libraries for almost all of the major tech stacks.

Reply:
I will add this to the tech spec under `Yet another message broker` section!

Reply:
Added the section.